FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/DemoTasks/DefenderDemoExample.c

```diff
@@ -0,0 +1,684 @@
+
+/**
+ * @brief Established connections array.
+ */
+static Connection_t pxEstablishedConnections[ democonfigESTABLISHED_CONNECTIONS_ARRAY_SIZE ];
+
+/**
+ * @brief All the metrics sent in the device defender report.
+ */
+static ReportMetrics_t xDeviceMetrics;
+
+/**
+ * @brief Report xStatus.
+ */
+static ReportStatus_t xReportStatus;
+
+/**
+ * @brief Buffer for generating the device defender report.
+ */
+static char pcDeviceMetricsJsonReport[ democonfigDEVICE_METRICS_REPORT_BUFFER_SIZE ];
+
+/**
+ * @brief Report Id sent in the defender report.
+ */
+static uint32_t ulReportId = 0;
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Callback to receive the incoming publish messages from the MQTT broker.
+ *
+ * @param[in] pPublishInfo Pointer to publish info of the incoming publish.
+ * @param[in] usPacketIdentifier Packet identifier of the incoming publish.
+ */
+static void prvPublishCallback( MQTTContext_t * pMqttContext,
+                                MQTTPacketInfo_t * pPacketInfo,
+                                MQTTDeserializedInfo_t * pDeserializedInfo );
+
+/**
+ * @brief Collect all the metrics to be sent in the device defender report.
+ *
+ * @return true if all the metrics are successfully collected;
+ * false otherwise.
+ */
+static bool prvCollectDeviceMetrics( void );
+
+/**
+ * @brief Generate the device defender report.
+ *
+ * @param[out] pulOutReportLength Length of the device defender report.
+ *
+ * @return true if the report is generated successfully;
+ * false otherwise.
+ */
+static bool prvGenerateDeviceMetricsReport( uint32_t * pulOutReportLength );
+
+/**
+ * @brief Subscribe to the device defender topics.
+ *
+ * @return true if the subscribe is successful;
+ * false otherwise.
+ */
+static bool prvSubscribeToDefenderTopics( void );
+
+/**
+ * @brief Unsubscribe from the device defender topics.
+ *
+ * @return true if the unsubscribe is successful;
+ * false otherwise.
+ */
+static bool prvUnsubscribeFromDefenderTopics( void );
+
+/**
+ * @brief Publish the generated device defender report.
+ *
+ * @param[in] ulReportLength Length of the device defender report.
+ *
+ * @return true if the report is published successfully;
+ * false otherwise.
+ */
+static bool prvPublishDeviceMetricsReport( uint32_t ulReportLength );
+
+/**
+ * @brief Validate the response received from the AWS IoT Device Defender Service.
+ *
+ * This functions checks that a valid JSON is received and the value of ulReportId
+ * is same as was sent in the published report.
+ *
+ * @param[in] pcDefenderResponse The defender response to validate.
+ * @param[in] ulDefenderResponseLength Length of the defender response.
+ *
+ * @return true if the response is valid;
+ * false otherwise.
+ */
+static bool prvValidateDefenderResponse( const char * pcDefenderResponse,
+                                         uint32_t ulDefenderResponseLength );
+
+/**
+ * @brief The task used to demonstrate the Defender API.
```

Comment:
Can you please add more details about what this task is doing in this documentation?